p2p/peer_behaviour.go

```diff
@@ -0,0 +1,137 @@
+package p2p
+
+import (
+	"errors"
+	"sync"
+)
+
+// Type of error behaviour a peer can perform.
+type ErrorPeerBehaviour int
+
+const (
+	ErrorPeerBehaviourUnknown = iota
+	ErrorPeerBehaviourBadMessage
+	ErrorPeerBehaviourMessageOutofOrder
+)
+
+// Type of good behaviour a peer can perform.
+type GoodPeerBehaviour int
+
+const (
+	GoodPeerBehaviourVote = iota + 100
+	GoodPeerBehaviourBlockPart
+)
+
+// PeerBehaviour provides an interface for reactors to signal the behaviour
+// of peers synchronously to other components.
+type PeerBehaviour interface {
+	Behaved(peerID ID, reason GoodPeerBehaviour) error
+	Errored(peerID ID, reason ErrorPeerBehaviour) error
+}
+
+type SwitchPeerBehaviour struct {
+	sw *Switch
+}
+
+// Reports the ErrorPeerBehaviour of peer identified by peerID to the Switch.
+func (spb *SwitchPeerBehaviour) Errored(peerID ID, reason ErrorPeerBehaviour) error {
+	peer := spb.sw.Peers().Get(peerID)
+	if peer == nil {
+		return errors.New("Peer not found")
+	}
+
+	spb.sw.StopPeerForError(peer, reason)
+	return nil
+}
+
+// Reports the GoodPeerBehaviour of peer identified by peerID to the Switch.
+func (spb *SwitchPeerBehaviour) Behaved(peerID ID, reason GoodPeerBehaviour) error {
+	peer := spb.sw.Peers().Get(peerID)
+	if peer == nil {
+		return errors.New("Peer not found")
+	}
+
+	spb.sw.MarkPeerAsGood(peer)
+	return nil
+}
+
+// Return a new switchPeerBehaviour instance which wraps the Switch.
+func NewSwitchPeerBehaviour(sw *Switch) *SwitchPeerBehaviour {
```

Comment:
This constructor should be moved above right after the type def of `SwitchPeerBehaviour`.